### PR TITLE
Add postgresql-hll extension

### PR DESCRIPTION
**Pull request by liuxueyang**
This Postgres module introduces a new data type hll which is a
HyperLogLog data structure. Refer to contrib/postgresql-hll/README.md
for more details.

Fix Makefile and unittest cases for Greenplum.

Add gpcontrib/README about how to use git subtree. We also use git
subtree for postgresql-hll.

Co-authored-by: Xuebin Su <sxuebin@vmware.com>

### CODE REVIEW

**Review by Aegeaner — APPROVED** [review]
LGTM.
Maybe better to use RANDOM distribution policy instead of the default HASH distribution policy to test the MPP case?

**Comment by higuoxing** [comment]
Greenplum has hyperloglog implementation as well, what's the difference between the original one and this extension? Can we remove the original implementation? https://github.com/greenplum-db/gpdb/blob/master/src/backend/utils/hyperloglog/gp_hyperloglog.c

**Comment by Aegeaner** [comment]
> Greenplum has hyperloglog implementation as well, what's the difference between the original one and this extension? Can we remove the original implementation? https://github.com/greenplum-db/gpdb/blob/master/src/backend/utils/hyperloglog/gp_hyperloglog.c

From the license info, I guess this is cherry-picked from an HLL implementation by CitusData: https://github.com/citusdata/postgresql-hll/  This implementation seems to have richer type support and even with libpq support.

But in the other hand, [the original GPDB implementation](https://github.com/greenplum-db/gpdb/blob/master/src/backend/utils/hyperloglog/gp_hyperloglog.c) applied several improvements that are described in ["HyperLogLog in Practice: Algorithmic Engineering of a State of The Art Cardinality Estimation Algorithm"](https://dl.acm.org/doi/10.1145/2452376.2452456), published by Stefan Heulem, Marc Nunkesse and Alexander Hall:
https://github.com/greenplum-db/gpdb/blob/fcddbd9d59f498ee97cd18851bfb2d4eccfab38e/src/backend/utils/hyperloglog/gp_hyperloglog.c#L345

But the CitusData's implementation is a naive and straightforward one:
https://github.com/citusdata/postgresql-hll/blob/774c58d860e74c314093d12144d1818b4534f613/src/hll.c#L2312

I am still not sure about the comparison of their efficiency by now.

File: contrib/postgresql-hll/src/hll.c [hunk 1]
```diff
@@ -0,0 +1,3866 @@
...
+{
+    int32 old_maxsparse = g_max_sparse;
+    int32 maxsparse = PG_GETARG_INT32(0);
+
+    if (maxsparse < -1)
+        ereport(ERROR,
+                (errcode(ERRCODE_DATA_EXCEPTION),
+                 errmsg("sparse threshold must be in range [-1,MAXINT]")));
+
+    g_max_sparse = maxsparse;
```
Comment (Sasasu) [inline contrib/postgresql-hll/src/hll.c:2833]: `hll_set_max_sparse()`  will modify some in process memory. if we want this UDF can take effect on all segments and coordinator we need to modify this UDF.

`SELECT x from gp_dist_random('gp_id');` only execute the UDF on segment, on coordinator the `g_max_spares` does not applied. to ensure the effect applied to all process. need run this UDF first then run it with `gp_dist_random()`

the other method is to modify this UDF, use `CdbDispatchCommand()` to send the SQL to all segments then execute the SQL on the coordinator.

or GPDB can guarantee a (buggy) session-level GUC transaction. change `g_max_sparse` to a GUC and use `SET g_max_sparse to 0` to broadcast the modification to all processes.

---

`g_max_sparse` is a optimize parameter, does not impact binary format. so the value can be different on segments.

any option above there is acceptable, but I think would better to add some `GPDB special notes` in `contrib/postgresql-hll/README.md`. mentioned `hll_set_max_sparse()` have a different behavior.

**Review by Sasasu — APPROVED** [review]

**Review by Sasasu — APPROVED** [review]

**Pull request closed without merging**